src/core/Class.js

```diff
@@ -8,110 +8,116 @@ import * as Util from './Util';
 
 // Thanks to John Resig and Dean Edwards for inspiration!
 
-export function Class() {}
-
-Class.extend = function (props) {
-
+export class Class {
 	// @function extend(props: Object): Function
 	// [Extends the current class](#class-inheritance) given the properties to be included.
 	// Returns a Javascript function that is a class constructor (to be called with `new`).
-	const NewClass = function (...args) {
+	static extend(props) {
+		class NewClass {
+			_initHooksCalled = false;
 
-		Util.setOptions(this);
+			constructor(...args) {
+				Util.setOptions(this);
 
-		// call the constructor
-		if (this.initialize) {
-			this.initialize.apply(this, args);
-		}
+				// call the constructor
+				if (this.initialize) {
+					this.initialize.apply(this, args);
+				}
+
+				// call all constructor hooks
+				this.callInitHooks();
+			}
 
-		// call all constructor hooks
-		this.callInitHooks();
-	};
+			callInitHooks() {
+				if (this._initHooksCalled) { return; }
 
-	const parentProto = this.prototype;
+				// collect all prototypes in chain
+				const protos = [];
+				let proto = Object.getPrototypeOf(this);
 
-	const proto = Object.create(parentProto);
-	proto.constructor = NewClass;
+				while (proto !== null) {
+					protos.push(proto);
+					proto = Object.getPrototypeOf(proto);
+				}
 
-	NewClass.prototype = proto;
+				// reverse so the parent prototype is first
+				protos.reverse();
 
-	// inherit parent's statics
-	for (const i in this) {
-		if (Object.hasOwn(this, i) && i !== 'prototype') {
-			NewClass[i] = this[i];
+				// call init hooks on each prototype
+				protos.forEach((proto) => {
+					const initHooks = proto._initHooks ?? [];
+					initHooks.forEach(hook => hook.call(this));
+				});
+
+				this._initHooksCalled = true;
+			}
 		}
-	}
 
-	// mix static properties into the class
-	if (props.statics) {
-		Util.extend(NewClass, props.statics);
-	}
+		const parentProto = this.prototype;
 
-	// mix includes into the prototype
-	if (props.includes) {
-		Util.extend.apply(null, [proto].concat(props.includes));
-	}
+		// Set the instance properties
+		Object.setPrototypeOf(NewClass.prototype, this.prototype);
 
-	// mix given properties into the prototype
-	Util.extend(proto, props);
-	delete proto.statics;
-	delete proto.includes;
+		// Hook up the static properties
+		Object.setPrototypeOf(NewClass, this);
 
-	// merge options
-	if (proto.options) {
-		proto.options = parentProto.options ? Object.create(parentProto.options) : {};
-		Util.extend(proto.options, props.options);
-	}
+		const proto = NewClass.prototype;
 
-	proto._initHooks = [];
+		// mix static properties into the class
+		if (props.statics) {
+			Util.extend(NewClass, props.statics);
+		}
 
-	// add method for calling all hooks
-	proto.callInitHooks = function () {
+		// mix includes into the prototype
+		if (props.includes) {
+			Util.extend.apply(null, [proto].concat(props.includes));

```

Comment:
Nit: rest args where possible?
```suggestion
			Util.extend(proto, ...props.includes);

```

Reply:
Haven't found a functional way to modernize this, perhaps we leave it as is. WDYT @mourner?

Reply:
Yeah, let's leave as is if it's not an easy fix.

Reply:
Ok, I will consider this resolved then.